eta/core/datasets.py

```diff
@@ -1338,3 +1375,842 @@ def required(cls):
+                helper_list.append(helper)
+
+        return helper_list
+
+    def _to_helper_list_video_objects(self, records):
+        '''Balancer._to_helper_list for object attributes in videos'''
+        helper_list = []
+
+        for i, record in enumerate(records):
+            labels = record.get_labels()
+            NO_ID = 'NO_ID'
+            helper_dict = defaultdict(set)
+
+            for frame_no, frame in labels.frames.items():
+                if (frame_no < record.clip_start_frame
+                        or frame_no >= record.clip_end_frame):
+                    continue
+
+                for detected_object in frame.objects:
+                    if detected_object.label != self.object_label:
+                        continue
+
+                    for attr in detected_object.attrs:
+                        if attr.name == self.attr_name:
+                            obj_idx = (
+                                detected_object.index
+                                if detected_object.index is not None
+                                else NO_ID
+                            )
+
+                            helper_dict[obj_idx].add(attr.value)
+
+                            break
+
+            # At this point, the keys of helper dict are unique
+            # object indices for objects of type self.object_label.
+            # The values are unique attribute values for self.attr_name.
+
+            if len(helper_dict):
+                helper = (i, [])
+                for s in helper_dict.values():
+                    helper[1].extend(s)
+                helper_list.append(helper)
+
+        return helper_list
+
+    def _get_target_count(self, counts):
+        '''Compute the target count that we would like to balance each value to.
+
+        Args:
+            A: the occurrence matrix computed in Balancer._get_occurrence_matrix
+
+        Returns: Integer target value
+        '''
+        if self.target_count:
+            return self.target_count
+        return int(np.quantile(counts, self.target_quantile))
+
+    def _get_keep_idxs(self, A, counts, target_count):
+        '''Algorithm fun! This function chooses the set of records to keep (and
+        remove).
+
+        There's still plenty of potential for testing and improvement here.
+
+
+        This problem can be posed as:
+            minimize:
+                ||np.dot(A, x) - b||
+            subject to:
+                x[i] is an element of [0, 1]
+
+        and different algorithms may be substituted in.
+
+        Args:
+            A: the occurrence matrix computed in Balancer._get_occurrence_matrix
+            counts (vector): the original counts for each value
+            target_count (int): the target value
+
+        Returns:
+            a list of integer indices to keep
+        '''
+        b = counts - target_count
+
+        x = self._greedy(A, b)
+
+        return np.where(x == 0)[0]
+
+    def _random(self, A, b):
```

Comment:
@tylerganter what about refactoring these functions into a configurable argument of the `Balancer` constructor? You could support a `method="random" or "greedy"` argument that allows the user to choose which of these methods they want to use. Or, you could allow them to provide a `method_fcn` that implements their own `(A, b, score_fcn) => x` method

Reply:
^this would allow these functions to be defined as top-level functions, not as private methods of this class, which is cleaner I think (they are interchangeable bits, not intimately tied to the class)

Reply:
As is, this is a private method that is never used, which is basically like including commented code in a PR (discouraged). Upgrading to a top-level method makes it publicly available, so it is now constructive code :)

Reply:
I made the algorithm a configurable in the initialization. I feel the algorithm is too inherently tied to balancing for me to want to expose them as non-private methods.

@brimoor the intention, which is something that may be a discussion point in regards to ETA convention, is to make it easy to reference if someone wanted to replace `greedy` with something else down the road. I don't think you would ever choose `random` over `greedy`, but it is a very simple algorithm that is in turn easy to understand and reference.